src/database.rs

```diff
@@ -152,6 +154,230 @@ pub fn current_revision<Db: ?Sized + Database>(db: &Db) -> Revision {
+        }
+    }
+
+    #[derive(serde::Deserialize)]
+    #[serde(field_identifier, rename_all = "lowercase")]
+    enum DatabaseField {
+        Runtime,
+        Ingredients,
+    }
+
+    pub struct DeserializeDatabase<'db>(pub &'db mut Zalsa);
+
+    impl<'de> de::DeserializeSeed<'de> for DeserializeDatabase<'_> {
+        type Value = ();
+
+        fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+        where
+            D: de::Deserializer<'de>,
+        {
+            // Note that we have to deserialize using a manual visitor here because the
+            // `Deserialize` derive does not support fields that use `DeserializeSeed`.
+            deserializer.deserialize_struct("Database", &["runtime", "ingredients"], self)
+        }
+    }
+
+    impl<'de> serde::de::Visitor<'de> for DeserializeDatabase<'_> {
+        type Value = ();
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("struct Database")
+        }
+
+        fn visit_map<V>(self, mut map: V) -> Result<(), V::Error>
+        where
+            V: serde::de::MapAccess<'de>,
+        {
+            let mut runtime = None;
+            let mut ingredients = None;
+
+            while let Some(key) = map.next_key()? {
+                match key {
+                    DatabaseField::Runtime => {
+                        if runtime.is_some() {
+                            return Err(serde::de::Error::duplicate_field("runtime"));
+                        }
+
+                        runtime = Some(map.next_value()?);
+                    }
+                    DatabaseField::Ingredients => {
+                        if ingredients.is_some() {
+                            return Err(serde::de::Error::duplicate_field("ingredients"));
+                        }
+
+                        ingredients = Some(map.next_value_seed(DeserializeIngredients(self.0))?);
+                    }
+                }
+            }
+
+            let mut runtime = runtime.ok_or_else(|| serde::de::Error::missing_field("runtime"))?;
+            let () = ingredients.ok_or_else(|| serde::de::Error::missing_field("ingredients"))?;
+
+            self.0.runtime_mut().deserialize_from(&mut runtime);
+
+            Ok(())
+        }
+    }
+
+    struct DeserializeIngredients<'db>(&'db mut Zalsa);
+
+    impl<'de> serde::de::Visitor<'de> for DeserializeIngredients<'_> {
+        type Value = ();
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("a map")
+        }
+
+        fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+        where
+            M: serde::de::MapAccess<'de>,
+        {
+            let DeserializeIngredients(zalsa) = self;
+
+            while let Some(name) = access.next_key::<&str>()? {
+                // TODO: Serialize the ingredient index directly.
```

Comment:
Technically the debug name is not a unique identifier, but I found it more helpful for debugging than the ingredient index. It could also just be a field on each ingredient.

Reply:
A separate (maybe even optinal) field would make sense to me.

Reply:
I removed the debug name for now, it's unfortunately very annoying to add extra fields because serde's macros aren't compatible with `DeserializeSeed` (which we need for some fields, because we deserialize into the table directly).